triopg/_triopg.py

```diff
@@ -121,6 +125,42 @@ async def prepare(self, *args, **kwargs):
         )
         return TrioStatementProxy(asyncpg_statement)
 
+    @asynccontextmanager
+    async def listen(self, channel, max_buffer_size):
+        """LISTEN on `channel` notifications and return memory channel to iterate over
+
+        max_buffer_size - memory channel max buffer size
+
+        For example:
+
+        async with conn.listen('some.changes', max_buffer_size=1) as notifications:
+            async for notification in notifications:
+                if notification != NOTIFY_OVERFLOW:
+                    print('Postgres notification received:', notification)
+        """
+
+        assert max_buffer_size >= 1
+        send_channel, receive_channel = trio.open_memory_channel(
+            max_buffer_size + 1
+        )
+
+        def _listen_callback(c, pid, chan, payload):
+            stats = send_channel.statistics()
+            if stats.current_buffer_used == stats.max_buffer_size - 1:
+                send_channel.send_nowait(NOTIFY_OVERFLOW)
+            try:
+                send_channel.send_nowait(payload)
+            except trio.WouldBlock:
+                pass  # drop payload on the floor
+
+        async with receive_channel, send_channel:
+            await self.add_listener(channel, _listen_callback)
+            try:
+                yield receive_channel
+            finally:
+                with trio.CancelScope(shield=True):
+                    await self.remove_listener(channel, _listen_callback)
```

Comment:
@touilleMan  Should we wrap `.remove_listener` with `@_shielded` instead of using `with trio.CancelScope(shield=True)` here?

Reply:
I think we shouldn't wrap it with `@_shielded`. It was weird to see how `remove_listener()` still went ahead in the face of cancellation. I wrote a test to assert the current behavior: bb748f1.